crypto-service/src/main/kotlin/net/corda/cryptoservice/CryptoService.kt

```
@@ -0,0 +1,38 @@
+package net.corda.cryptoservice
+
+import org.bouncycastle.operator.ContentSigner
+import java.security.KeyPair
+import java.security.PublicKey
+
+interface CryptoService {
+
+    /**
+     * Generate and store a new [KeyPair].
+     * Note that schemeNumberID is Corda specific. Cross-check with the network operator for supported schemeNumberID
+     * and their corresponding signature schemes. The main reason for using schemeNumberID and not algorithm OIDs is
+     * because some schemes might not be standardised and thus an official OID might for this scheme not exist yet.
+     *
+     * Returns the [PublicKey] of the generated [KeyPair].
+     */
+    fun generateKeyPair(alias: String, schemeNumberID: Int): PublicKey
+
+    /** Check if this [CryptoService] has a private key entry for the input alias. */
+    fun containsKey(alias: String): Boolean
+
+    /**
+     * Returns the [PublicKey] of the input alias or null if it doesn't exist.
+     */
+    fun getPublicKey(alias: String): PublicKey?
+
+    /**
+     * Sign a [ByteArray] using the private key identified by the input alias.
+     * @return the signature bytes whose format depends on the underlying signature scheme and it should
+     * be Java BouncyCastle compatible (i.e., ASN.1 DER-encoded for ECDSA).
+     */
+    fun sign(alias: String, data: ByteArray): ByteArray
```

Using plain old file-based KeyStores, it is reasonable to assume that this operation will succeed in a timely manner. Now, we have different failure modes. Signing can take significantly longer, due to e.g. network latencies, signing might be (temporarily) not possible due to connectivity issues/ timeouts. Where and how do we address these problems best?
Example: We want to sign a transaction, but the (cloud-based) HSM of our node temporarily responds with 5xx. Do we want the signing operation to fail hard, should we retry with exponential back-off?
Another example: We want to sign a transaction, but the (cloud-based) HSM of our node is down, and does not even respond with an HTTP error code. Do we want to wait for a timeout to occur? Should we block until that happens?
@mikehearn what do you think?

---

Btw all key related methods in `KeyManagementService` are `@Suspendable` anyway. Defining a timeout policy is a different thing though. Wrt retry, IMO this should be implemented internally to each `CryptoServiceImpl` because there might be different requirements/guidelines per vendor (i.e., they can only accept certain requests per second - especially remote HSMs).